test/built-ins/FinalizationGroup/prototype/cleanupSome/iterator-dynamic.js

```diff
@@ -4,103 +4,54 @@
+     throw a TypeError exception.
 
-  8. If finalizationGroup.[[Cells]] contains a Record cell such that cell.[[Target]] is empty,
-    a. Choose any such cell.
-    b. Remove cell from finalizationGroup.[[Cells]].
-    c. Return CreateIterResultObject(cell.[[Holdings]], false).
-  9. Otherwise, return CreateIterResultObject(undefined, true).
-features: [FinalizationGroup, host-gc-required, Symbol]
+features: [FinalizationGroup, host-gc-required]
 includes: [async-gc.js]
 flags: [async, non-deterministic]
 ---*/
 
 var called = 0;
 var endOfCall = 0;
-var first, second, third;
-var firstIter, secondIter, thirdIter;
+var exception = undefined;
 var fg = new FinalizationGroup(function() {});
 
 function callback(iterator) {
   called += 1;
 
   if (called === 1) {
-
-    // Inception!
-    fg.cleanupSome(callback);
-
-    first = iterator.next();
-    firstIter = iterator;
-  } else if (called === 2) {
-
-    // Double Inception!
-    fg.cleanupSome(callback);
-
-    second = iterator.next();
-    secondIter = iterator;
-  } else if (called === 3) {
-
-    // Triple Inception!
-    // But this time we won't try to consume iterator anymore, leave it alone there.
-    fg.cleanupSome(callback);
-
-    third = iterator.next();
-    thirdIter = iterator;
+    // Atempt to re-enter the callback.
+    try {
+      fg.cleanupSome(callback);
+    } catch (e) {
+      exception = e;
+    }
   }
 
   endOfCall += 1;
 }
 
 function emptyCells() {
   var o1 = {};
-  var o2 = {};
   fg.register(o1, 'holdings 1');
-  fg.register(o2, 'holdings 2');
 
-  var prom = asyncGC(o1, o2);
+  var prom = asyncGC(o1);
   o1 = null;
-  o2 = null;
 
   return prom;
 }
 
 emptyCells().then(function() {
   fg.cleanupSome(callback);
 
-  // Make sure everything is set
-  assert.sameValue(called, 4, 'cleanup successfully');
-  assert.sameValue(endOfCall, 4, 'cleanup ended successfully');
-  
-  assert.notSameValue(firstIter, secondIter, 'callback is not called with the same iterator #1');
-  assert.notSameValue(firstIter, thirdIter, 'callback is not called with the same iterator #2');
-  assert.notSameValue(secondIter, thirdIter, 'callback is not called with the same iterator #3');
-  
-  assert.sameValue(first.value, undefined, 'iterator is already consumed');
-  assert.sameValue(first.done, true, 'first callback will find no empty Targets');
-  assert.sameValue(second.done, false, 'second callback will find an empty Target');
-  assert.sameValue(third.done, false, 'third callback will find an empty Target');
-  
-  // 8.a. Choose any such cell.
-  var holdings = [second.value, third.value];
-  
-  assert(holdings.includes('holdings 1'), 'iterators consume emptied cells #1');
-  assert(holdings.includes('holdings 2'), 'iterators consume emptied cells #2');  
+  assert.sameValue(called, 1, 'callback was called');
+  assert.sameValue(endOfCall, 1, 'callback finished');
+  assert(exception instanceof TypeError);
```

Comment:
This test also removes one very important test: The fact you can have multiple simultaneous iterators and consume them in different orders...

so now you can't call cleanupSome while `IsFinalizationGroupCleanupJobActive === true` but you can call consecutive cleanupSomes and store each iterator, than consume and verify the order.

So the iterator remains dynamic. I suggest having 2 test files.

1. Observe the TypeError of calling cleanupSome inside a cleanupSome callback (what this PR change is doing)
  - Also observe the same of 1, but after the iterator is fully consumed so the instance is clean, but still throws a TypeError...
2. Observe dynamic consumption of iterators for the same fg instance



```js
// 1
function callback(iterator) {
  called += 1;
  var otherCallbacks = 0;
  assert.throws(TypeError, () => {
    fg.cleanupSome(() => { otherCallbacks += 1; });
  });

  [...iterator];
  assert.throws(TypeError, () => {
    fg.cleanupSome(() => { otherCallbacks += 1; });
  });
  assert.sameValue(otherCallbacks, 0, 'other callbacks are never called');
  endOfCall += 1;
}

fg.cleanUpSome(callback);
```

```js
// 2
var iterators = [];
function callback(iterator) {
  iterators.push(iterator);
}

fg.cleanUpSome(callback);
fg.cleanUpSome(callback);

assert.sameValue(iterators.length, 2);

var [ firstIter, secondIter ] = iterators;

var first = firstIter.next();

fg.cleanUpSome(callback);
assert.sameValue(iterators.length, 3);
var thirdIter = iterators[2];

assert.notSameValue(firstIter, secondIter, 'callback is not called with the same iterator #1');
assert.notSameValue(firstIter, thirdIter, 'callback is not called with the same iterator #2');
assert.notSameValue(secondIter, thirdIter, 'callback is not called with the same iterator #3');

var second = secondIter.next();
var third = thirdIter.next();

assert.sameValue(third.value, undefined, 'iterator is already consumed');
assert.sameValue(third.done, true, 'no empty Targets');

assert.sameValue(first.done, false, 'first found an empty Target');
assert.sameValue(second.done, false, 'second found an empty Target');

fg.cleanUpSome(callback);
assert.sameValue(iterators.length, 3, 'callback is not called again with no empty targets');

// assert holdings here
...
```

Reply:
Actually because of https://github.com/tc39/proposal-weakrefs/issues/159 you can't use an iterator outside of its cleanup callback anymore. Combined with https://github.com/tc39/proposal-weakrefs/issues/158, there is no way to create 2 concurrently active iterators.